presto-main/src/main/java/com/facebook/presto/operator/exchange/LocalExchange.java

```diff
@@ -0,0 +1,584 @@
+    {
+        return sources.get(partitionIndex);
+    }
+
+    private void sourceFinished()
+    {
+        checkNotHoldsLock(this);
+
+        if (!sources.stream().allMatch(LocalExchangeSource::isFinished)) {
+            return;
+        }
+
+        // all sources are finished, so finish the sinks
+        ImmutableList<LocalExchangeSink> openSinks;
+        synchronized (this) {
+            allSourcesFinished = true;
+
+            // since all sources are finished there is no reason to all new pages
+            openSinks = ImmutableList.copyOf(sinks);
+            sinks.clear();
+        }
+
+        openSinks.forEach(LocalExchangeSink::finished);
+        checkAllSinksComplete();
+    }
+
+    private LocalExchangeSink createSink(LocalExchangeSinkFactory factory)
+    {
+        checkNotHoldsLock(this);
+
+        // Note: exchanger can be stateful so create a new one for each sink
+        LocalExchangeSink sink = new LocalExchangeSink(this, exchangerSupplier.get());
+        synchronized (this) {
+            checkState(sinkFactories.contains(factory), "Factory is already closed");
+
+            if (!allSourcesFinished) {
+                sinks.add(sink);
+                return sink;
+            }
+        }
+
+        // all sources have already finished so return a finished sink
+        // NOTE: finish must be called outside of the lock since it calls back
+        sink.finished();
+        return sink;
+    }
+
+    private void sinkFinished(LocalExchangeSink sink)
+    {
+        checkNotHoldsLock(this);
+
+        synchronized (this) {
+            sinks.remove(sink);
+        }
+        checkAllSinksComplete();
+    }
+
+    private void noMoreSinkFactories()
+    {
+        checkNotHoldsLock(this);
+
+        synchronized (this) {
+            noMoreSinkFactories = true;
+        }
+        checkAllSinksComplete();
+    }
+
+    private void sinkFactoryClosed(LocalExchangeSinkFactory sinkFactory)
+    {
+        checkNotHoldsLock(this);
+
+        synchronized (this) {
+            sinkFactories.remove(sinkFactory);
+        }
+        checkAllSinksComplete();
+    }
+
+    private void checkAllSinksComplete()
+    {
+        checkNotHoldsLock(this);
+
+        synchronized (this) {
+            if (!noMoreSinkFactories || !sinkFactories.isEmpty() || !sinks.isEmpty()) {
+                return;
+            }
+        }
+
+        sources.forEach(LocalExchangeSource::finish);
+        writeWaiter.finishAndNotifyWaiters();
+    }
+
+    private ListenableFuture<?> waitForWriting()
+    {
+        checkNotHoldsLock(this);
+
+        synchronized (this) {
+            if (bufferBytes.get() < maxBufferedBytes) {
+                return NOT_BLOCKED;
+            }
+        }
+        return writeWaiter.waitFor();
+    }
+
+    private static void checkNotHoldsLock(Object lock)
+    {
+        checkState(!Thread.holdsLock(lock), "Can not execute this method while holding a lock");
+    }
+
+    @ThreadSafe
```

Comment:
The thread-safety of this class seems complicated to me.

`removePage` does not acquire lock at all while all other operations involving `buffer`, `bufferedBytes` and `finishing` seem to provide atomicity. As a result, for example, `onFinish` may be invoked multiple times. 

Can you state the guarantees provided and not provided here?